### PR TITLE
fix #31674, error when storing nonzeros into structural zeros with .=

### LINKED ISSUE #31674: Structured array .= silently ignores unstructured non-zeros
Consider the following:
```
julia> U = UpperTriangular(Float64[1 2 3; 0 1 2; 0 0 1])
3×3 UpperTriangular{Float64,Array{Float64,2}}:
 1.0  2.0  3.0
  ⋅   1.0  2.0
  ⋅    ⋅   1.0

julia> X = randn(3, 3)
3×3 Array{Float64,2}:
  0.9189    -0.942656  -0.825467
 -0.346463   0.829813   0.741945
 -0.239974  -0.232412   1.46933 

julia> U .+= X
3×3 UpperTriangular{Float64,Array{Float64,2}}:
 1.9189  1.05734  2.17453
  ⋅      1.82981  2.74194
  ⋅       ⋅       2.46933
```
Note that `U` was updated with the result of elementwise addition with `X`. However, `U` is still an `UpperTriangular`, even though the right hand side of the addition is not. Generally speaking, the result of adding two matrices is only upper triangular if both are.

It seems to me that the case of updating a structured matrix should not work except when the operation would produce a matrix of the same structure.

### PR DESCRIPTION
**Pull request by mbauman**
Previously, broadcasted assignment (`.=`) would happily ignore all nonstructured portions of the destination, regardless of whether the broadcasted expression would actually evaluate to zero or not. This changes these in-place methods to use the same infrastructure that out-of-place broadcast uses to determine the result type. If we are unsure of the structural properties of the output, we fall back to the generic implementation, which will attempt to store into every single location of the destination -- including those structural zeros. Thus we now error in cases where we generate nonzeros in those locations.

### CODE REVIEW

File: stdlib/LinearAlgebra/test/structuredbroadcast.jl [hunk 1]
```diff
@@ -51,14 +51,37 @@ end
     A = rand(N, N)
     sA = A + copy(A')
     D = Diagonal(rand(N))
-    B = Bidiagonal(rand(N), rand(N - 1), :U)
+    Bu = Bidiagonal(rand(N), rand(N - 1), :U)
+    Bl = Bidiagonal(rand(N), rand(N - 1), :L)
     T = Tridiagonal(rand(N - 1), rand(N), rand(N - 1))
+    ◣ = LowerTriangular(rand(N,N))
```
Comment (ararslan) [inline stdlib/LinearAlgebra/test/structuredbroadcast.jl]: :heart:

**Comment by andreasnoack** [comment]
Why is it necessary to add special checks for this instead? I would assume that, in general, in-place broadcasting would automatically fail when there are structural zeros with only whitelisted special cases (zero preserving mapings) would work.

**Comment by mbauman** [comment]
Yes, the general broadcast methods just do the right thing.  An alternative PR would be to simply remove all these `copyto!` methods.

But these `copyto!` methods are huge optimizations!  They only store into the structural nonzeros (typically O(n)) and completely ignore all the calculations of all the zeros (typically O(n^2)).  Of course we can only perform these optimizations when its safe.  Thus the checks.  If a given broadcast expression is not known to be safe, then we punt out to the general broadcast method (which is of course just O(n^2)).

**Comment by JeffBezanson** [comment]
Should this be merged?

**Comment by mbauman** [comment]
Yes please!  This is good by me; I'm probably just overly conservative about waiting for express approval before merging my own PRs.

**Pull request merged**